index.rst

```diff
@@ -1,29 +1,605 @@
+Messages are sent using TCP/IP.
+
+Both DDS and Kafka support optionally retrieving historical (also called "late-joiner") data, and our control system architecture requires this.
+As a node starts up, it reads the most recent historical message for each event it subscribes to, in order to get the current state of the system without having to issue commands to CSCs.
+However, we do not read historical data for commands, because reponding to stale commands is dangerous.
+We also do not read historical telemetry because telemetry is output frequently, so there is no need, and we want to be sure it the data is current.
+
+Differences Between DDS and Kafka
+=================================
+
+Kafka is a broker-based system, whereas DDS is distributed, with no central broker.
+
+With DDS each node (e.g. CSC or Jupyter Notebook) discovers the other nodes, using UDP multicast, as the software starts up.
+Reliability is achieved by switching which node acts as the primary source of data if one fails.
+DDS (using our current configuration) publishes topic data using multicast, which means each message is only sent over the network once.
+
+With Kafka each node connects to a specified broker and schema registry; there is no discovery phase and no need for UDP multicast.
+Kafka writers write messages to the broker, and readers read messages from the broker.
+This is a much simpler topology than DDS, but it does require that each message be sent to the broker and then again to each subscriber.
+Reliability and performance can be increased by running multiple brokers in parallel, and by using more than one partition for each topic.
+
+Schema Evolution
+----------------
+
+DDS does not support schema evolution, at least as we are using it.
+Thus, in order to prevent collisions between old topic schemas and new topic schemas, the DDS topic name includes a hash that we change whenever the schema changes.
+Our high-level software hides that hash from users.
+
+Kafka schemas support basic schema evolution: we may add or delete fields and readers and writers can freely use older or newer schemas without restarting.
+Furthermore, even if a schema is not compatible for some topics, all we need to do is restart the publishers and subscribers for those topics to use the new schema; we do not have to bring down the whole system, as we do for DDS.
+
+Licensing
+---------
+
+With ADLink OpenSplice we presently run two very different versions of the software:
+
+* Developers use the free community edition of OpenSplice, in order to reduce our costs and to avoid licensing hassles.
+  ADLink no longer offers a free edition of OpenSplice, and it only ever supported single-process topic daemons, which cannot handle the demands of our full control system.
+  The free edition was intentionally kept one major version behind the licensed version; when it was withdrawn it was at version 6.9, while the licensed software was at 6.10.
```

Comment:
Wasn't the licensed version already at 6.11 when the community edition got pulled?